### PR TITLE
Convert `@classkernel` to `@kernel`

### PR DESCRIPTION
**Pull request by k-ye**
Related issue id = #548 

Basically, we used the trick in https://stackoverflow.com/a/8793684/12003165 to detect if a function being decorated belongs to a class or not. Note that we cannot apply that trick as-is -- To support OOP, Taichi classes *must* be decorated by `@ti.data_oriented`. So instead of looking for `class `, we look for `@(\w+\.)?data_oriented`

Another thing being unified is `grad` on the wrapped kernel. There is no need for the special kwarg `_gradient: bool` on class kernels now. I made a small change to `DifferentiableMethod` (renamed to  `BoundedDifferentiableMethods`) so that it can get `adjoint` from the passed in wrapped function.

Also, both types of kernels' primal `Kernel` instance have a `grad` referencing the adjoint instance. In this way we are also able to unify the way to insert `(primal, args)` to a `Tape`.

**Limits**

* Users must decorate their classes with `@ti.data_oriented`. They still have the freedom to name `taichi` however they want (not limited to `ti`), but they must not alias the name `data_oriented`.

### CODE REVIEW

**Comment by k-ye** [comment]
I should point out a case that doesn't pass even on `master` branch: Having more than one instance of a Taichi class + tape usage.

```python
@ti.host_arch
def test_oop_two_items():

  @ti.data_oriented
  class Array2D:

    def __init__(self, n, m, increment, multiplier):
      self.n = n
      self.m = m
      self.val = ti.var(ti.f32)
      self.total = ti.var(ti.f32)
      self.increment = increment
      self.multiplier = multiplier

    def place(self, root):
      root.dense(ti.ij, (self.n, self.m)).place(self.val)
      root.place(self.total)

    @ti.kernel
    def inc(self):
      for i, j in self.val:
        ti.atomic_add(self.val[i, j], self.increment)

    @ti.kernel
    def reduce(self):
      for i, j in self.val:
        ti.atomic_add(self.total, self.val[i, j] * self.multiplier)

  arr1_inc, arr1_mult = 3, 4
  arr2_inc, arr2_mult = 6, 8
  arr1 = Array2D(128, 128, arr1_inc, arr1_mult)
  arr2 = Array2D(16, 32, arr2_inc, arr2_mult)

  @ti.layout
  def place():
    # Place an object. Make sure you defined place for that obj
    ti.root.place(arr1)
    ti.root.place(arr2)
    ti.root.lazy_grad()

  with ti.Tape(loss=arr1.total):
    arr1.reduce()
  with ti.Tape(loss=arr2.total):
    arr2.reduce()

  for i in range(arr1.n):
    for j in range(arr1.m):
      assert arr1.val.grad[i, j] == arr1_mult
  for i in range(arr2.n):
    for j in range(arr2.m):
      assert arr2.val.grad[i, j] == arr2_mult
```

If the two `with ti.Tape()` calls are invoked before checking `grad[,]`, then `arr1.val.grad` is zero... However, if I write it this way:

```python
  with ti.Tape(loss=arr1.total):
    arr1.reduce()
  for i in range(arr1.n):
    for j in range(arr1.m):
      assert arr1.val.grad[i, j] == arr1_mult

  with ti.Tape(loss=arr2.total):
    arr2.reduce()
  for i in range(arr2.n):
    for j in range(arr2.m):
      assert arr2.val.grad[i, j] == arr2_mult
```

Then things worked. Is this a bug in Taichi OOP or just an illegal usage of `Tape`?

**Comment by yuanming-hu** [comment]
This is a really, really great new feature to have. Users do not have to distinguish `ti.kernel` and `ti.classkernel` anymore. Thank you so much!

> Then things worked. Is this a bug in Taichi OOP or just an illegal usage of Tape?

`Tape` should ideally work with no issue here. I suspect it to be a Taichi OOP issue.

**Comment by yuanming-hu** [comment]
Oh I know what's wrong. It's the `Tape` that clears all the gradients.

**Comment by k-ye** [comment]
Ahh! Let me try again

**Comment by yuanming-hu** [comment]
This is actually the desired behavior of `Tape`. To get what you want, just use
```python
...
with ti.Tape(loss=arr2.total, clear_gradients=False):
    arr2.reduce()
...
```

Maybe we should update the doc. This is indeed confusing :-)

**Comment by k-ye** [comment]
> This is actually the desired behavior of `Tape`. To get what you want, just use
> 
> ```python
> ...
> with ti.Tape(loss=arr2.total, clear_gradients=False):
>     arr2.reduce()
> ...
> ```
> 
> Maybe we should update the doc. This is indeed confusing :-)

Cool! It's working on this PR as well XD

**Comment by k-ye** [comment]
btw, do you think it's simpler to rename `ti.data_oriented` to `ti.oop`..?

**Comment by yuanming-hu** [comment]
I agree with you that `ti.data_oriented` is a bit too long. `ti.oop` might be slightly unclear and sounds like `oops` - how about `ti.object`?

Btw, apart from
```
@ti.object
class A:
   ...
```
Maybe we can also make use of metaclasses
```
class A(ti.object):
   ...
```

**Comment by k-ye** [comment]
LOL. I feel like `ti.object` is a bit too generic (and may be confused with python's `object`). 

Name aside, I do like the Metaclass approach -- that way we can remove `_DECORATED_CLASS_STACKFRAME_STMT_RE` and just check for `class `, which makes the approach a bit more reliable.  But on the other hand, i don't know if it would be easier for users to forget to inherit from this metaclass... Or we need to find a way inside @ti.kernel, to check if the class of that method inherits from the metaclass? Not really sure if that's possible, because the class definition hasn't completed by the time @ti.kernel executes..

**Comment by k-ye** [comment]
Looks like Python3.8 changed its stack layout when interpreting class definitions....

**Comment by yuanming-hu** [comment]
Let's stick to `ti.data_oriented` for now, since it reminds people that they are doing data-oriented programming, in an (object-oriented) class. As an experienced Taichi programmer, I do like `ti.oop`, but it looks slightly less informative to newcomers.

I'm always open to new decorator name ideas though, if we come up with one that is significantly better :-)

**Comment by k-ye** [comment]
Sounds good, that suggestion was really just a random thought w/o much consideration... We have much bigger things to think through for now XD

**Comment by ppwwyyxx** [comment]
btw, there is another uncommon way to define classes (which is what `class A: ...` is transformed into):
```
def a_method(instance):
    print(instance.x)

A = type('A', (), {'a_method': a_method, 'x': 1})
print(A().a_method())
```

There is no perfect way to tell whether a function will be used as a method or not at the time the function is defined. 
Customization can be done when class is defined (through a metaclass) or immediately after class is defined (through class decorator `ti.data_oriented`). Since both happen after the function (and the `Kernel` object) is defined, using them would require something like `kernel.extract_arguments_again(classkernel=True)` that recreates arguments & `KernelTemplateMapper` after knowing the kernel will used as a method.

**Comment by k-ye** [comment]
> Since both happen after the function (and the Kernel object) is defined, using them to achieve the same goal would require something like kernel.extract_arguments_again(classkernel=True)

Ah OK, I guess in another word, this means to delay the initialization of `Kernel` to when we customize the class, but not when the function is defined? Thanks for pointing this out!
